File: wasm/sample/src/xplpc/client/proxy-client.ts
import { XPLPC } from "../core/xplpc";
import { XMappingList } from "../data/mapping-list";
import { XMessage } from "../message/message";
import { ExceptionMessage } from "../util/exception-message";
import { Log } from "../util/log";

export class XProxyClient {
    static call(data: string): Promise<string> {
        return new Promise<string>((resolve) => {
            if (!XPLPC.shared().initialized) {
                Log.e("[XProxyClient : call] The WASM module is not initialized");
                return resolve("");
            }

            // function name
            const functionName = XPLPC.shared().config.serializer.decodeFunctionName(data)

            if (!functionName) {
                Log.e("[ProxyClient : call] Function name is empty")
                return resolve("");
            }

            // mapping item
            const mappingItem = XMappingList.shared().find(functionName)

            if (!mappingItem) {
                Log.e("[ProxyClient : call] Mapping not found for function: " + functionName)
                return resolve("");
            }

            // execute
            let message: XMessage | undefined = undefined;

            try {
                message = XPLPC.shared().config.serializer.decodeMessage(data)
            } catch (e: unknown) {
                Log.e("[ProxyClient : call] Error when decode message: " + ExceptionMessage.get(e))
            }

            if (!message) {
                Log.e("[ProxyClient : call] Error when decode message for function: " + functionName)
                return resolve("");
            }

            try {
                return resolve(mappingItem.target(message).then((response: unknown) => {
                    return XPLPC.shared().config.serializer.encodeFunctionReturnValue(response)
                }));
            } catch (e: unknown) {
                Log.e("[ProxyClient : call] Error when encode message: " + ExceptionMessage.get(e))
                return resolve("");
            }
        });
    }
}

File: flutter/plugin/lib/client/proxy_client.dart
import 'package:xplpc/core/xplpc.dart';
import 'package:xplpc/data/mapping_list.dart';
import 'package:xplpc/message/message.dart';
import 'package:xplpc/type/typedefs.dart';
import 'package:xplpc/util/log.dart';

class ProxyClient {
  static void call(String data, ProxyClientCallback callback) {
    // function name
    var functionName = XPLPC.instance.config.serializer.decodeFunctionName(
      data,
    );

    if (functionName.isEmpty) {
      Log.e("[ProxyClient : call] Function name is empty");
      callback?.call("");
      return;
    }

    // mapping item
    var mappingItem = MappingList.instance.find(functionName);

    if (mappingItem == null) {
      Log.e(
        "[ProxyClient : call] Mapping not found for function: $functionName",
      );
      callback?.call("");
      return;
    }

    // execute
    Message? message;

    try {
      message = XPLPC.instance.config.serializer.decodeMessage(data);
    } catch (e) {
      Log.e("[ProxyClient : call] Error when decode message: $e");
    }

    if (message == null) {
      Log.e(
        "[ProxyClient : call] Error when decode message for function: $functionName",
      );
      callback?.call("");
      return;
    }

    try {
      mappingItem.target(message, (dynamic r) {
        callback
            ?.call(XPLPC.instance.config.serializer.encodeFunctionReturnValue(
          r,
        ));
      });

      return;
    } catch (e) {
      Log.e("[ProxyClient : call] Error: $e");
    }

    callback?.call("");
  }
}
